File: src/java/org/apache/fop/util/ColorWithFallback.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.util;

import java.awt.Color;
import java.awt.color.ColorSpace;

import org.apache.xmlgraphics.java2d.color.ColorWithAlternatives;

/**
 * This class is a {@link Color} subclass adding a fallback color that FOP uses to re-serialize
 * color specifications as textual functions. The fallback is otherwise not used in producing
 * output formats.
 */
public class ColorWithFallback extends ColorWithAlternatives {

    private static final long serialVersionUID = 7913922854959637136L;

    private final Color fallback;

    /**
     * Creates a new color
     * @param cspace the color space of the primary color
     * @param components the color components
     * @param alpha the alpha component
     * @param alternativeColors the array of alternative colors if applicable (may be null)
     * @param fallback the fallback color (usually an sRGB color)
     */
    public ColorWithFallback(ColorSpace cspace, float[] components, float alpha,
            Color[] alternativeColors, Color fallback) {
        super(cspace, components, alpha, alternativeColors);
        this.fallback = fallback;
    }

    /**
     * Copy constructor adding a fallback color.
     * @param color the color to be duplicated
     * @param fallback the fallback color (usually an sRGB color)
     */
    public ColorWithFallback(Color color, Color fallback) {
        this(color.getColorSpace(), color.getColorComponents(null),
                getAlphaFloat(color), getAlternativeColors(color), fallback);
    }

    private static float getAlphaFloat(Color color) {
        float[] comps = color.getComponents(null);
        return comps[comps.length - 1]; //Alpha is on last component
    }

    private static Color[] getAlternativeColors(Color color) {
        if (color instanceof ColorWithAlternatives) {
            ColorWithAlternatives cwa = (ColorWithAlternatives)color;
            if (cwa.hasAlternativeColors()) {
                return cwa.getAlternativeColors();
            }
        }
        return null;
    }

    /**
     * Returns the fallback color.
     * @return the fallback color
     */
    public Color getFallbackColor() {
        return this.fallback;
    }

}

File: dotnet/src/Fop.Colors/ColorWithFallback.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

namespace Fop.Colors;

/// <summary>
/// A <see cref="FopColor"/> carrying a fallback colour that FOP uses to
/// re-serialize colour specifications as textual functions. The fallback is
/// otherwise not used in producing output formats.
/// <para>
/// Port of <c>org.apache.fop.util.ColorWithFallback</c>. In Java this extended
/// <c>ColorWithAlternatives</c>; in this slice of the port the alternate
/// colour-space representation (space name + native components) is carried
/// directly on <see cref="FopColor"/>, so this type only adds the fallback.
/// </para>
/// </summary>
public sealed class ColorWithFallback : FopColor
{
    private readonly string? _colorSpaceName;
    private readonly float[] _components;
    private readonly FopColor _fallback;

    /// <summary>
    /// Creates a new colour with an alternate colour space and a fallback.
    /// The packed sRGB/alpha value of this colour is taken from the fallback so
    /// that the red/green/blue accessors return usable sRGB approximations.
    /// </summary>
    /// <param name="colorSpaceName">the name of the primary colour space (e.g. a
    /// pseudo-profile such as <c>#CMYK</c> or <c>#Separation</c>)</param>
    /// <param name="components">the native colour components for that space</param>
    /// <param name="fallback">the fallback colour (usually an sRGB colour)</param>
    public ColorWithFallback(string? colorSpaceName, float[] components, FopColor fallback)
        : base(fallback.Rgb)
    {
        ArgumentNullException.ThrowIfNull(components);
        ArgumentNullException.ThrowIfNull(fallback);
        _colorSpaceName = colorSpaceName;
        _components = (float[])components.Clone();
        _fallback = fallback;
    }

    /// <summary>The fallback colour (usually an sRGB colour).</summary>
    public FopColor FallbackColor => _fallback;

    /// <inheritdoc/>
    public override string? ColorSpaceName => _colorSpaceName;

    /// <inheritdoc/>
    public override IReadOnlyList<float> Components => _components;

    // A ColorWithFallback carries its alternate colour space directly; it does
    // not also carry a separate array of alternative Color objects.
    /// <inheritdoc/>
    public override bool HasAlternativeColors => false;

    /// <inheritdoc/>
    public override bool Equals(FopColor? other)
    {
        if (other is not ColorWithFallback cwf)
        {
            return false;
        }

        if (ReferenceEquals(this, other))
        {
            return true;
        }

        return Rgb == cwf.Rgb
            && _colorSpaceName == cwf._colorSpaceName
            && _components.AsSpan().SequenceEqual(cwf._components)
            && _fallback.Equals(cwf._fallback);
    }

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(Rgb, _colorSpaceName, _fallback);
}
